# Attr data cleaning

In [1]:
import pandas as pd
import numpy as np

In [2]:
dat = pd.read_csv("dataset_used/attr.csv")
data = dat.copy()
pd.set_option('display.max_columns', None)

In [3]:
data.sample(5)

id  Age Attrition     BusinessTravel  DailyRate  \
1279  1280   44       Yes  Travel_Frequently        429   
599    600   36        No      Travel_Rarely       1041   
1201  1202   23       Yes      Travel_Rarely       1320   
540    541   28       Yes      Travel_Rarely        654   
1040  1041   40        No         Non-Travel        218   

                  Department  DistanceFromHome  Education   EducationField  \
1279  Research & Development                 1          2          Medical   
599          Human Resources                13          3  Human Resources   
1201  Research & Development                 8          1          Medical   
540   Research & Development                 1          2    Life Sciences   
1040  Research & Development                 8          1          Medical   

      EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  \
1279              1            1792                        3    Male   
599               1             829                        3    Male   
1201              1            1684                        4    Male   
540               1             741                        1  Female   
1040              1            1468                        4    Male   

      HourlyRate  JobInvolvement  JobLevel                JobRole  \
1279          99               3         1     Research Scientist   
599           36               3         1        Human Resources   
1201          93               2         1  Laboratory Technician   
540           67               1         1     Research Scientist   
1040          55               2         3      Research Director   

      JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
1279                2      Divorced           2342        11092   
599                 2       Married           2143        25527   
1201                3        Single           3989        20586   
540                 2        Single           2216         3872   
1040                2      Divorced          13757        25178   

      NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
1279                   1      Y      Yes                 12   
599                    4      Y       No                 13   
1201                   1      Y      Yes                 11   
540                    7      Y      Yes                 13   
1040                   2      Y       No                 11   

      PerformanceRating  RelationshipSatisfaction  StandardHours  \
1279                  3                         3             80   
599                   3                         2             80   
1201                  3                         1             80   
540                   3                         4             80   
1040                  3                         3             80   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
1279                 3                  6                      2   
599                  1                  8                      2   
1201                 0                  5                      2   
540                  0                 10                      4   
1040                 1                 16                      5   

      WorkLifeBalance  YearsAtCompany  YearsInCurrentRole  \
1279                2               5                   3   
599                 3               5                   2   
1201                3               5                   4   
540                 3               7                   7   
1040                3               9                   8   

      YearsSinceLastPromotion  YearsWithCurrManager  
1279                        2                     3  
599                         0                     4  
1201                        1                     2  
540                         3                     7  
1040                        4                     8

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 36 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   id                        1470 non-null   int64 
 1   Age                       1470 non-null   int64 
 2   Attrition                 1470 non-null   object
 3   BusinessTravel            1470 non-null   object
 4   DailyRate                 1470 non-null   int64 
 5   Department                1470 non-null   object
 6   DistanceFromHome          1470 non-null   int64 
 7   Education                 1470 non-null   int64 
 8   EducationField            1470 non-null   object
 9   EmployeeCount             1470 non-null   int64 
 10  EmployeeNumber            1470 non-null   int64 
 11  EnvironmentSatisfaction   1470 non-null   int64 
 12  Gender                    1470 non-null   object
 13  HourlyRate                1470 non-null   int64 
 14  JobInvolvement          

The `data.info()` allows us to see a quick overview of our data and with that we were able to deduce that the data has no missing values and it also has 36 columns and 1470 rows of data, with each column being either an integer ot an object

# Cleaning the data column by column

### Age

For this we can create a bin of ages but for that we would have to get the youngest and oldest age in the column so we know how to create the bin

In [5]:
for i in sorted(data.Age.unique()):
    print(i, end=", ")

18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 

In [6]:
print(f"The youngest age in the data is {data.Age.unique().min()}\n\
The oldest age in the data is {data.Age.unique().max()}")

The youngest age in the data is 18
The oldest age in the data is 60


In [7]:
bins = [18, 20, 30, 40, 50, 60]
group = ['<20', '20-29', '30-39', '40-49', '50-59']

# Filling NaN with 60+ because we used the right False keyword which will by default ignore the rightmost digit (60)
data['AgeGroup'] = pd.cut(data['Age'], bins=bins, labels=group, right=False).cat.add_categories('60+').fillna('60+')

### Attrition

In [8]:
data.Attrition.unique()

array(['Yes', 'No'], dtype=object)

Everything looks okay here so we move on to the next
### BusinessTravel

In [9]:
data.BusinessTravel.unique()

array(['Travel_Rarely', 'Travel_Frequently', 'Non-Travel'], dtype=object)

In [10]:
data.BusinessTravel = data.BusinessTravel.str.replace('_', ' ').str.replace("-", " ")
data.BusinessTravel.unique()

array(['Travel Rarely', 'Travel Frequently', 'Non Travel'], dtype=object)

Everything looks okay here so we move on to the next
### Department

In [11]:
data.Department.unique()

array(['Sales', 'Research & Development', 'Human Resources'], dtype=object)

Everything looks okay here so we move on to the next
### DistanceFromHome

In [12]:
for i in sorted(data.DistanceFromHome.unique()):
    print(i, end=', ')

1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 

In [13]:
# I will create a bin for the distance
bins = [1, 11, 20, 30]
labels = ['Very Close', 'Close', 'Far']
data['DistanceFrom_Home'] = pd.cut(data['DistanceFromHome'], bins=bins, labels=labels, right=False)

Everything looks okay here so we move on to the next
### Education

In [14]:
data.Education.unique()

array([2, 1, 4, 3, 5], dtype=int64)

In [15]:
data.Education =  data.Education.map({1: "Below College",
                                      2: "College",
                                      3: "Bachelor",
                                      4: "Master",
                                      5: "Doctor"})

In [16]:
data.Education.unique()

array(['College', 'Below College', 'Master', 'Bachelor', 'Doctor'],
      dtype=object)

Everything looks okay here so we move on to the next
### EducationField

In [17]:
data.EducationField.unique()

array(['Life Sciences', 'Other', 'Medical', 'Marketing',
       'Technical Degree', 'Human Resources'], dtype=object)

Everything looks okay here so we move on to the next
### EmployeeCount

In [18]:
data.EmployeeCount.unique()

array([1], dtype=int64)

Everything looks okay here so we move on to the next
### EnvironmentSatisfaction

In [19]:
data.EnvironmentSatisfaction.unique()

array([2, 3, 4, 1], dtype=int64)

In [20]:
data.EnvironmentSatisfaction = data.EnvironmentSatisfaction.map({1: 'Low',
                                                                 2: 'Medium',
                                                                 3: 'High',
                                                                 4: 'Very High'})

In [21]:
data.EnvironmentSatisfaction.unique()

array(['Medium', 'High', 'Very High', 'Low'], dtype=object)

Everything looks okay here so we move on to the next
### Gender

In [22]:
data.Gender.unique()

array(['Female', 'Male'], dtype=object)

Everything looks okay here so we move on to the next
### JobInvolvement

In [23]:
data.JobInvolvement.unique()

array([3, 2, 4, 1], dtype=int64)

In [24]:
data.JobInvolvement = data.JobInvolvement.map({1: 'Low',
                                               2: 'Medium',
                                               3: 'High',
                                               4: 'Very High'})

In [25]:
data.JobInvolvement.unique()

array(['High', 'Medium', 'Very High', 'Low'], dtype=object)

Everything looks okay here so we move on to the next
### JobLevel

In [26]:
data.JobLevel.unique()

array([2, 1, 3, 4, 5], dtype=int64)

In [27]:
data.JobLevel = "Lvl " + data.JobLevel.astype(str)

In [28]:
data.JobLevel.unique()

array(['Lvl 2', 'Lvl 1', 'Lvl 3', 'Lvl 4', 'Lvl 5'], dtype=object)

Everything looks okay here so we move on to the next
### JobRole

In [29]:
data.JobRole.unique()

array(['Sales Executive', 'Research Scientist', 'Laboratory Technician',
       'Manufacturing Director', 'Healthcare Representative', 'Manager',
       'Sales Representative', 'Research Director', 'Human Resources'],
      dtype=object)

Everything looks okay here so we move on to the next
### JobSatisfaction

In [30]:
data.JobSatisfaction.unique()

array([4, 2, 3, 1], dtype=int64)

In [31]:
data.JobSatisfaction =data.JobSatisfaction.map({1: 'Low',
                                                2: 'Medium',
                                                3: 'High',
                                                4: 'Very High'})

In [32]:
data.JobSatisfaction.unique()

array(['Very High', 'Medium', 'High', 'Low'], dtype=object)

Everything looks okay here so we move on to the next
### MaritalStatus

In [33]:
data.MaritalStatus.unique()

array(['Single', 'Married', 'Divorced'], dtype=object)

Everything looks okay here so we move on to the next
### OverTime

In [34]:
data.OverTime.unique()

array(['Yes', 'No'], dtype=object)

Everything looks okay here so we move on to the next
### PerformanceRating

In [35]:
data.PerformanceRating.unique()

array([3, 4], dtype=int64)

In [36]:
data.PerformanceRating = data.PerformanceRating.map({3: "Excellent",
                                                     4: "Outstanding"})

In [37]:
data.PerformanceRating.unique()

array(['Excellent', 'Outstanding'], dtype=object)

Everything looks okay here so we move on to the next
### RelationshipSatisfaction

In [38]:
data.RelationshipSatisfaction.unique()

array([1, 4, 2, 3], dtype=int64)

In [39]:
data.RelationshipSatisfaction = data.RelationshipSatisfaction.map({1: 'Low',
                                                                   2: 'Medium',
                                                                   3: 'High',
                                                                   4: 'Very High'})

In [40]:
data.RelationshipSatisfaction.unique()

array(['Low', 'Very High', 'Medium', 'High'], dtype=object)

Everything looks okay here so we move on to the next
### StockOptionLevel

In [41]:
data.StockOptionLevel.unique()

array([0, 1, 3, 2], dtype=int64)

In [42]:
data.StockOptionLevel = "Lvl " + data.StockOptionLevel.astype(str)

In [43]:
data.StockOptionLevel.unique()

array(['Lvl 0', 'Lvl 1', 'Lvl 3', 'Lvl 2'], dtype=object)

Everything looks okay here so we move on to the next
### WorkLifeBalance

In [44]:
data.WorkLifeBalance.unique()

array([1, 3, 2, 4], dtype=int64)

In [45]:
data.WorkLifeBalance = data.WorkLifeBalance.map({1: 'Bad',
                                                 2: 'Good',
                                                 3: 'Better',
                                                 4: 'Best'})

In [46]:
data.WorkLifeBalance.unique()

array(['Bad', 'Better', 'Good', 'Best'], dtype=object)

Everything looks okay here so we move on to the next
### All columns that deals with years

In [47]:
# We start by creating BINS for this year group
bins = [0, 10, 20, 30]
group = ['0-10 years', '11-20 years', '21-29 years']

# Creating BINS
data['TotalWorkingYears'] = pd.cut(data['TotalWorkingYears'],
                                   bins=bins,
                                   labels=group,
                                   right=False).cat.add_categories('31-40 years').fillna('31-40 years')


data['YearsAtCompany'] = pd.cut(data['YearsAtCompany'],
                                bins=bins,
                                labels=group,
                                right=False).cat.add_categories('31-40 years').fillna('31-40 years')

In [48]:
# Appending years to the remainig data that deals with years
yrs = ['TrainingTimesLastYear', 'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']

In [49]:
# Funciton to append years or year to data
def years(col):
    data[col] = data[col].astype(str) + " " + np.where(data[col] == 0, 'year', 'years')

In [50]:
for i in yrs:
    years(i)
    print(f"{i} Successfully imputed\n")

TrainingTimesLastYear Successfully imputed

YearsInCurrentRole Successfully imputed

YearsSinceLastPromotion Successfully imputed

YearsWithCurrManager Successfully imputed



In [51]:
data.sample(4)

id  Age Attrition     BusinessTravel  DailyRate  \
1151  1152   28        No      Travel Rarely        580   
629    630   28        No      Travel Rarely       1169   
446    447   41        No         Non Travel        267   
584    585   42        No  Travel Frequently        570   

                  Department  DistanceFromHome Education EducationField  \
1151  Research & Development                27  Bachelor        Medical   
629          Human Resources                 8   College        Medical   
446                    Sales                10   College  Life Sciences   
584   Research & Development                 8  Bachelor  Life Sciences   

      EmployeeCount  EmployeeNumber EnvironmentSatisfaction  Gender  \
1151              1            1622                  Medium  Female   
629               1             869                  Medium    Male   
446               1             599               Very High    Male   
584               1             809                  Medium    Male   

      HourlyRate JobInvolvement JobLevel                 JobRole  \
1151          39            Low    Lvl 2  Manufacturing Director   
629           63         Medium    Lvl 1         Human Resources   
446           56           High    Lvl 2         Sales Executive   
584           66           High    Lvl 5                 Manager   

     JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
1151             Low      Divorced           4877        20460   
629        Very High      Divorced           4936        23965   
446        Very High        Single           6230        13430   
584        Very High      Divorced          18430        16225   

      NumCompaniesWorked Over18 OverTime  PercentSalaryHike PerformanceRating  \
1151                   0      Y       No                 21       Outstanding   
629                    1      Y       No                 13         Excellent   
446                    7      Y       No                 14         Excellent   
584                    1      Y       No                 13         Excellent   

     RelationshipSatisfaction  StandardHours StockOptionLevel  \
1151                   Medium             80            Lvl 1   
629                 Very High             80            Lvl 1   
446                 Very High             80            Lvl 0   
584                    Medium             80            Lvl 1   

     TotalWorkingYears TrainingTimesLastYear WorkLifeBalance YearsAtCompany  \
1151        0-10 years               5 years            Good     0-10 years   
629         0-10 years               6 years          Better     0-10 years   
446        11-20 years               3 years          Better    11-20 years   
584        21-29 years               4 years            Good    21-29 years   

     YearsInCurrentRole YearsSinceLastPromotion YearsWithCurrManager AgeGroup  \
1151            3 years                  0 year               0 year    20-29   
629             1 years                  0 year              4 years    20-29   
446             3 years                 1 years             10 years    40-49   
584             7 years                14 years              9 years    40-49   

     DistanceFrom_Home  
1151               Far  
629         Very Close  
446         Very Close  
584         Very Close

#### Selecting the columns that would bring more meaning during analysis

In [52]:
data = data[['AgeGroup', 'Gender', 'MaritalStatus', 'Attrition', 'BusinessTravel', 'DistanceFrom_Home', 'Department',
             'Education', 'EducationField', 'EnvironmentSatisfaction', 'JobInvolvement', 'JobLevel', 'JobRole',
             'JobSatisfaction', 'HourlyRate', 'DailyRate', 'MonthlyIncome', 'MonthlyRate', 'NumCompaniesWorked',
             'OverTime', 'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction', 'StandardHours',
             'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance', 'YearsAtCompany', 
             'YearsInCurrentRole', 'YearsSinceLastPromotion', 'YearsWithCurrManager']]

In [53]:
# Old dataframe
dat.sample(10)

id  Age Attrition     BusinessTravel  DailyRate  \
467    468   37        No         Non-Travel        142   
502    503   53        No      Travel_Rarely        238   
1266  1267   41        No      Travel_Rarely        548   
438    439   35        No      Travel_Rarely       1276   
977    978   34        No         Non-Travel        999   
1233  1234   30        No      Travel_Rarely        793   
205    206   29       Yes      Travel_Rarely        121   
332    333   54        No  Travel_Frequently        928   
705    706   39        No      Travel_Rarely        903   
1283  1284   28        No      Travel_Rarely       1181   

                  Department  DistanceFromHome  Education    EducationField  \
467                    Sales                 9          4           Medical   
502                    Sales                 1          1           Medical   
1266  Research & Development                 9          4     Life Sciences   
438   Research & Development                16          3     Life Sciences   
977   Research & Development                26          1  Technical Degree   
1233  Research & Development                16          1     Life Sciences   
205                    Sales                27          3         Marketing   
332   Research & Development                20          4     Life Sciences   
705                    Sales                 2          5     Life Sciences   
1283  Research & Development                 1          3     Life Sciences   

      EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  \
467               1             626                        1    Male   
502               1             682                        4  Female   
1266              1            1772                        3    Male   
438               1             586                        4    Male   
977               1            1374                        1  Female   
1233              1            1729                        2    Male   
205               1             283                        2  Female   
332               1             450                        4  Female   
705               1             985                        1    Male   
1283              1            1799                        3    Male   

      HourlyRate  JobInvolvement  JobLevel                    JobRole  \
467           69               3         3            Sales Executive   
502           34               3         2            Sales Executive   
1266          94               3         1      Laboratory Technician   
438           72               3         3  Healthcare Representative   
977           92               2         1         Research Scientist   
1233          33               3         1         Research Scientist   
205           35               3         3            Sales Executive   
332           31               3         2         Research Scientist   
705           41               4         3            Sales Executive   
1283          82               3         1         Research Scientist   

      JobSatisfaction MaritalStatus  MonthlyIncome  MonthlyRate  \
467                 2      Divorced           8834        24666   
502                 1        Single           8381         7507   
1266                1      Divorced           2289        20520   
438                 3       Married           7632        14295   
977                 3      Divorced           2029        15891   
1233                4       Married           2862         3811   
205                 4       Married           7639        24525   
332                 3        Single           4869        16885   
705                 3        Single           7880         2560   
1283                4       Married           2044         5531   

      NumCompaniesWorked Over18 OverTime  PercentSalaryHike  \
467                    1      Y       No                 13   
502                    7   

In [54]:
# New dataframe
data.sample(10)

AgeGroup  Gender MaritalStatus Attrition BusinessTravel  \
719     40-49  Female        Single        No  Travel Rarely   
1058    30-39  Female        Single       Yes  Travel Rarely   
770     40-49    Male      Divorced        No  Travel Rarely   
1216    40-49    Male       Married        No  Travel Rarely   
928     40-49  Female       Married       Yes  Travel Rarely   
29      40-49  Female        Single        No  Travel Rarely   
1151    20-29  Female      Divorced        No  Travel Rarely   
1349    20-29  Female       Married        No  Travel Rarely   
774     50-59    Male        Single        No     Non Travel   
1080    40-49  Female       Married        No  Travel Rarely   

     DistanceFrom_Home              Department      Education EducationField  \
719         Very Close                   Sales        College  Life Sciences   
1058               Far                   Sales         Master        Medical   
770         Very Close  Research & Development         Master        Medical   
1216        Very Close                   Sales       Bachelor        Medical   
928              Close  Research & Development       Bachelor        Medical   
29          Very Close                   Sales         Master      Marketing   
1151               Far  Research & Development       Bachelor        Medical   
1349        Very Close  Research & Development        College  Life Sciences   
774         Very Close  Research & Development  Below College        Medical   
1080        Very Close                   Sales       Bachelor  Life Sciences   

     EnvironmentSatisfaction JobInvolvement JobLevel  \
719                Very High           High    Lvl 2   
1058                     Low         Medium    Lvl 2   
770                Very High           High    Lvl 5   
1216               Very High           High    Lvl 2   
928                      Low           High    Lvl 3   
29                    Medium           High    Lvl 5   
1151                  Medium            Low    Lvl 2   
1349                  Medium         Medium    Lvl 1   
774                     High         Medium    Lvl 4   
1080                    High           High    Lvl 4   

                        JobRole JobSatisfaction  HourlyRate  DailyRate  \
719             Sales Executive       Very High          83        955   
1058            Sales Executive          Medium          40        790   
770           Research Director       Very High          40        430   
1216            Sales Executive       Very High          73       1179   
928   Healthcare Representative       Very High          73        621   
29                      Manager             Low          83        705   
1151     Manufacturing Director             Low          39        580   
1349         Research Scientist            High          90        482   
774                     Manager             Low          40        444   
1080                    Manager          Medium          51        228   

      MonthlyIncome  MonthlyRate  NumCompaniesWorked OverTime  \
719            4163         8571                   1      Yes   
1058           4599         7815                   0      Yes   
770           19627        21445                   9       No   
1216           7847         6069                   1      Yes   
928            7978        14075                   1       No   
29            18947        22822                   3       No   
1151           4877        20460                   0       No   
1349           2933        14908                   1      Yes   
774           16756        17323                   7       No   
1080          16606        11380                   8       No   

      PercentSalaryHike PerformanceRating RelationshipSatisfaction  \
719                  17         Excellent                     High   
1058                 23       Outstanding                     High   
770                  17         Excellent                Very

In [55]:
# Save to csv
data.to_csv("attr_cleaned.csv")

# The end